state-transition/core/state_processor_validators.go

```diff
@@ -21,53 +21,147 @@
+			}
+		}
 	}
 
-	// pick prev epoch validators
+	// validators registry will be possibly further modified in order to enforce
+	// validators set cap. We will do that at the end of processEpoch, once all
+	// Eth 2.0 like transitions has been done (notable EffectiveBalances handling).
+	return nil
+}
+
+//nolint:lll // let it be
+func (sp *StateProcessor[
+	_, _, _, BeaconStateT, _, _, _, _, _, _, _, _, ValidatorT, _, _, _, _,
+]) processValidatorSetCap(
+	st BeaconStateT,
+) error {
+	// Enforce the validator set cap by:
+	// 1- retrieving validators active next epoch
+	// 2- sorting them by stake
+	// 3- dropping enough validators to fulfill the cap
+
 	slot, err := st.GetSlot()
 	if err != nil {
-		return nil, err
+		return err
 	}
+	nextEpoch := sp.cs.SlotToEpoch(slot) + 1
 
-	sp.valSetMu.Lock()
-	defer sp.valSetMu.Unlock()
+	nextEpochVals, err := sp.getActiveVals(st, nextEpoch)
+	if err != nil {
+		return fmt.Errorf("registry update, failed retrieving next epoch vals: %w", err)
+	}
 
-	// prevEpoch is calculated assuming current block
-	// will turn epoch but we have not update slot yet
-	prevEpoch := sp.cs.SlotToEpoch(slot)
-	currEpoch := prevEpoch + 1
-	if slot == 0 {
-		currEpoch = 0 // prevEpoch for genesis is zero
+	if uint64(len(nextEpochVals)) <= sp.cs.ValidatorSetCap() {
+		// nothing to eject
+		return nil
 	}
-	prevEpochVals := sp.valSetByEpoch[prevEpoch] // picks nil if it's genesis
 
-	// calculate diff
-	res := sp.validatorSetsDiffs(prevEpochVals, activeVals)
+	slices.SortFunc(nextEpochVals, func(lhs, rhs ValidatorT) int {
+		var (
+			val1Stake = lhs.GetEffectiveBalance()
+			val2Stake = rhs.GetEffectiveBalance()
+		)
+		switch {
+		case val1Stake < val2Stake:
+			return -1
+		case val1Stake > val2Stake:
+			return 1
+		default:
+			// validators pks are guaranteed to be different
+			var (
+				val1Pk = lhs.GetPubkey()
+				val2Pk = rhs.GetPubkey()
+			)
+			return stdbytes.Compare(val1Pk[:], val2Pk[:])
+		}
+	})
 
-	// clear up sets we won't lookup to anymore
-	sp.valSetByEpoch[currEpoch] = activeVals
-	if prevEpoch >= 1 {
-		delete(sp.valSetByEpoch, prevEpoch-1)
+	// We do not currently have a cap on validators churn, so we stop
+	// validators next epoch and we withdraw them the epoch after
+	var idx math.ValidatorIndex
+	for li := range uint64(len(nextEpochVals)) - sp.cs.ValidatorSetCap() {
+		valToEject := nextEpochVals[li]
+		valToEject.SetExitEpoch(nextEpoch)
+		valToEject.SetWithdrawableEpoch(nextEpoch + 1)
+		idx, err = st.ValidatorIndexByPubkey(valToEject.GetPubkey())
+		if err != nil {
+			return fmt.Errorf("validators cap, failed loading validator index: %w", err)
+		}
+		if err = st.UpdateValidatorAtIndex(idx, valToEject); err != nil {
+			return fmt.Errorf("validator cap, failed ejecting validator idx %d: %w", li, err)
+		}
 	}
-	return res, nil
+
+	return nil
 }
 
 // Note: validatorSetsDiffs does not need to be a StateProcessor method
```

Comment:
can we add a TODO here to remove the StateProcessor receiver here once the generics removals are done by @fridrik01? Just so we can make this a stateless helper function asap.

Reply:
Done